// terraform-generator/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const TerraformGenerator_1 = __importDefault(require("./TerraformGenerator"));
exports.default = TerraformGenerator_1.default;
var Block_1 = require("./blocks/Block");
exports.Block = Block_1.default;
var Provider_1 = require("./blocks/Provider");
exports.Provider = Provider_1.default;
var Resource_1 = require("./blocks/Resource");
exports.Resource = Resource_1.default;
var Data_1 = require("./blocks/Data");
exports.Data = Data_1.default;
var Output_1 = require("./blocks/Output");
exports.Output = Output_1.default;
var Module_1 = require("./blocks/Module");
exports.Module = Module_1.default;
var Variable_1 = require("./blocks/Variable");
exports.Variable = Variable_1.default;
var Backend_1 = require("./blocks/Backend");
exports.Backend = Backend_1.default;
var Argument_1 = require("./arguments/Argument");
exports.Argument = Argument_1.default;
exports.arg = Argument_1.arg;
var Attribute_1 = require("./arguments/Attribute");
exports.Attribute = Attribute_1.default;
exports.attr = Attribute_1.attr;
var Heredoc_1 = require("./arguments/Heredoc");
exports.Heredoc = Heredoc_1.default;
exports.heredoc = Heredoc_1.heredoc;
var Function_1 = require("./arguments/Function");
exports.Function = Function_1.default;
exports.fn = Function_1.fn;
var Map_1 = require("./types/Map");
exports.Map = Map_1.default;
exports.map = Map_1.map;
