// terraform-generator/dist/blocks/Resource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const __1 = require("..");
class Resource extends __1.Block {
    /**
     * Construct resource.
     * Refer to Terraform documentation on what can be put as type & arguments.
     *
     * @param type type
     * @param name name
     * @param args arguments
     */
    constructor(type, name, args) {
        super('resource', [type, name], args);
        this.type = type;
        this.name = name;
    }
    asArgument() {
        return new __1.Argument(`${this.type}.${this.name}`);
    }
    attr(name) {
        return new __1.Attribute(this, name);
    }
    /**
     * Convert resource into data source.
     * Refer to Terraform documentation on what can be put as arguments.
     *
     * @param options options
     * @param argNames names of resource arguments to converted into data source arguments;
     * use array for name mapping, position 0 = original resource's argument name, position 1 = mapped data source's argument name
     * @param args extra arguments
     */
    toData(options, argNames, args) {
        const type = (options && options.type) ? options.type : this.type;
        const name = (options && options.name) ? options.name : this.name;
        if (!args) {
            args = {};
        }
        if (!args['filter']) {
            args['filter'] = [];
        }
        else if (!Array.isArray(args['filter'])) {
            throw new Error('Filter is not an array.');
        }
        for (const argName of argNames) {
            let actualArgName;
            let newArgName;
            if (typeof argName === 'string') {
                actualArgName = argName;
                newArgName = argName;
            }
            else {
                actualArgName = argName[0];
                newArgName = argName[1];
            }
            const arg = this.getArgument(actualArgName);
            if (arg instanceof __1.Map) {
                for (const mapArgName in arg.args) {
                    args['filter'].push({
                        name: `${newArgName}:${mapArgName}`,
                        values: [arg.args[mapArgName]]
                    });
                }
            }
            else if (!args[newArgName]) {
                args[newArgName] = arg;
            }
        }
        return new __1.Data(type, name, args);
    }
}
exports.default = Resource;
